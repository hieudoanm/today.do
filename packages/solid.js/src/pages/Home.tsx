import { createSignal, onMount, For, Show } from 'solid-js';
import { Task, createIndexedDB } from '~/signals/createIndexedDB';

export const Home = () => {
	const { database, list, add, update, remove } = createIndexedDB<Task>({
		databaseName: 'TasksDB',
		storeName: 'tasks',
	});

	const [tasks, setTasks] = createSignal<Task[]>([]);
	const [newTask, setNewTask] = createSignal('');

	// Load tasks once DB is ready
	onMount(async () => {
		const db = database();
		if (!db) {
			// Watch for when DB initializes
			const checkDB = setInterval(async () => {
				if (database()) {
					clearInterval(checkDB);
					const items = await list();
					setTasks(items);
				}
			}, 100);
			return;
		}

		const items = await list();
		setTasks(items);
	});

	const handleAdd = async () => {
		const text = newTask().trim();
		if (!text) return;
		await add(text);
		setNewTask('');
		const updatedTasks = await list();
		setTasks(updatedTasks);
	};

	const handleToggle = async (todo: Task) => {
		await update({ ...todo, completed: !todo.completed });
		const updatedTasks = await list();
		setTasks(updatedTasks);
	};

	const handleDelete = async (id: number) => {
		await remove(id);
		const updatedTasks = await list();
		setTasks(updatedTasks);
	};

	return (
		<div class="bg-base-200 flex h-screen flex-col items-center py-0 md:min-h-screen md:py-10">
			<div class="bg-base-100 h-screen w-full max-w-md rounded-none p-6 shadow-lg md:h-auto md:rounded-2xl">
				<h1 class="mb-4 text-center text-2xl font-bold">Tasks</h1>

				{/* Input Section */}
				<div class="mb-4 flex gap-2">
					<input
						type="text"
						placeholder="Add a new task"
						class="input input-bordered flex-1"
						value={newTask()}
						onInput={(e) => setNewTask(e.currentTarget.value)}
						onKeyDown={(e) => e.key === 'Enter' && handleAdd()}
					/>
					<button class="btn btn-primary rounded-full" onClick={handleAdd}>
						Add
					</button>
				</div>

				{/* Task List */}
				<ul class="space-y-2">
					<For each={tasks()}>
						{(todo) => (
							<li class="bg-base-300 flex items-center justify-between rounded-lg p-3">
								<div class="flex items-center gap-2">
									<input
										type="checkbox"
										class="checkbox checkbox-primary"
										checked={todo.completed}
										onChange={[handleToggle, todo]}
									/>
									<span
										class={`text-lg ${todo.completed ? 'text-gray-500 line-through' : ''}`}>
										{todo.text}
									</span>
								</div>
								<button
									class="btn btn-sm btn-error rounded-full"
									onClick={[handleDelete, todo.id]}>
									Delete
								</button>
							</li>
						)}
					</For>
				</ul>

				<Show when={tasks().length === 0}>
					<p class="mt-4 text-center text-gray-500">No tasks yet!</p>
				</Show>
			</div>
		</div>
	);
};
