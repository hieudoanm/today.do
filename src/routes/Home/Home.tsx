import { Task, useIndexedDB } from '@today.do/hooks/useIndexedDB';
import { FC, useEffect, useState } from 'react';
import { PiPlus, PiX } from 'react-icons/pi';

export const Home: FC = () => {
	const { database, list, add, update, remove } = useIndexedDB<Task>({
		databaseName: 'TasksDB',
		storeName: 'tasks',
	});

	const [tasks, setTasks] = useState<Task[]>([]);
	const [newTask, setNewTask] = useState<string>('');

	useEffect(() => {
		if (!database) return;
		list().then(setTasks);
	}, [database]);

	const handleAdd = async () => {
		if (!newTask.trim()) return;
		await add(newTask.trim());
		setNewTask('');
		const updatedTask = await list();
		setTasks(updatedTask);
	};

	const handleToggle = async (todo: Task) => {
		await update({ ...todo, completed: !todo.completed });
		const updatedTask = await list();
		setTasks(updatedTask);
	};

	const handleDelete = async (id: number) => {
		await remove(id);
		const updatedTask = await list();
		setTasks(updatedTask);
	};

	return (
		<div className="bg-base-200 flex h-screen flex-col items-center py-0 md:min-h-screen md:py-10">
			<div className="bg-base-100 h-screen w-full max-w-md rounded-none p-6 shadow-lg md:h-auto md:rounded-2xl">
				<h1 className="mb-4 text-center text-2xl font-bold">Tasks</h1>

				{/* Input Section */}
				<div className="mb-4 flex gap-2">
					<input
						type="text"
						placeholder="Add a new task"
						className="input input-bordered flex-1"
						value={newTask}
						onChange={(e) => setNewTask(e.target.value)}
						onKeyDown={(e) => e.key === 'Enter' && handleAdd()}
					/>
					<button className="btn btn-glass" onClick={handleAdd}>
						<PiPlus />
					</button>
				</div>

				{/* Task List */}
				<ul className="space-y-2">
					{tasks.map((todo) => (
						<li key={todo.id} className="bg-base-300 flex items-center justify-between rounded-lg p-3">
							<div className="flex items-center gap-2">
								<input
									type="checkbox"
									className="checkbox checkbox-primary"
									checked={todo.completed}
									onChange={() => handleToggle(todo)}
								/>
								<span className={`text-lg ${todo.completed ? 'text-gray-500 line-through' : ''}`}>{todo.text}</span>
							</div>
							<button className="btn btn-sm btn-error" onClick={() => handleDelete(todo.id)}>
								<PiX />
							</button>
						</li>
					))}
				</ul>

				{tasks.length === 0 && <p className="mt-4 text-center text-gray-500">No tasks yet!</p>}
			</div>
		</div>
	);
};
