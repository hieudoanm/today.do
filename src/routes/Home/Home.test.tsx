import { render } from '@testing-library/react';
import { Home } from './Home';

describe('Home', () => {
	it('should match snapshot', () => {
		const { container } = render(<Home />);
		expect(container).toMatchSnapshot();
	});
});
