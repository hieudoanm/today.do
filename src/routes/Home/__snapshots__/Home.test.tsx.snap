// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home should match snapshot 1`] = `
<div>
  <div
    class="bg-base-200 flex h-screen flex-col items-center py-0 md:min-h-screen md:py-10"
  >
    <div
      class="bg-base-100 h-screen w-full max-w-md rounded-none p-6 shadow-lg md:h-auto md:rounded-2xl"
    >
      <h1
        class="mb-4 text-center text-2xl font-bold"
      >
        Tasks
      </h1>
      <div
        class="mb-4 flex gap-2"
      >
        <input
          class="input input-bordered flex-1"
          placeholder="Add a new task"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224,128a8,8,0,0,1-8,8H136v80a8,8,0,0,1-16,0V136H40a8,8,0,0,1,0-16h80V40a8,8,0,0,1,16,0v80h80A8,8,0,0,1,224,128Z"
            />
          </svg>
        </button>
      </div>
      <ul
        class="space-y-2"
      />
      <p
        class="mt-4 text-center text-gray-500"
      >
        No tasks yet!
      </p>
    </div>
  </div>
</div>
`;
